klippa/src/lib.rs

```diff
@@ -7,67 +7,216 @@ mod parsing_util;
+                    break;
+                }
+                self.glyphset_gsub.insert(gid);
             }
         }
+        self.glyphset_gsub
+            .extend(self.unicode_to_new_gid_list.iter().map(|t| t.1));
+        self.unicodes
+            .extend(self.unicode_to_new_gid_list.iter().map(|t| t.0));
+    }
+
+    pub fn populate_gids_to_retain(&mut self, font: &FontRef) {
+        //not-def
+        self.glyphset_gsub.insert(GlyphId::from(0));
+
+        //glyph closure for cmap
+        let cmap = font.cmap().expect("Error reading cmap table");
+        cmap.closure_glyphs(&self.unicodes, &mut self.glyphset_gsub);
+        remove_invalid_gids(&mut self.glyphset_gsub, self.font_num_glyphs);
+
+        //skip glyph closure for MATH table, it's not supported yet
+        //glyph closure for COLR
+        //TODO: colr_closure;
+        //remove_invalid_gids(glyphset_colr, font_num_glyphs);
+
+        /* Populate a full set of glyphs to retain by adding all referenced composite glyphs. */
+        let loca = font.loca(None).expect("Error reading loca table");
+        let glyf = font.glyf().expect("Error reading glyf table");
+        let operation_count =
+            self.glyphset_gsub.len() * (MAX_COMPOSITE_OPERATIONS_PER_GLYPH as usize);
+        for gid in self.glyphset_gsub.iter() {
+            glyf_closure_glyphs(
+                &loca,
+                &glyf,
+                gid,
+                &mut self.glyphset,
+                operation_count as i32,
+                0,
+            );
+        }
+        remove_invalid_gids(&mut self.glyphset, self.font_num_glyphs);
+    }
+}
+
+/// glyph closure for Composite glyphs in glyf table
+fn glyf_closure_glyphs(
+    loca: &Loca,
+    glyf: &Glyf,
+    gid: GlyphId,
+    gids_to_retain: &mut IntSet<GlyphId>,
+    operation_count: i32,
+    depth: u8,
+) -> i32 {
+    if gids_to_retain.contains(gid) {
+        return operation_count;
+    }
+    gids_to_retain.insert(gid);
+
+    if depth > MAX_NESTING_LEVEL {
+        return operation_count;
+    }
+    let depth = depth + 1;
+
+    let mut operation_count = operation_count - 1;
+    if operation_count < 0 {
+        return operation_count;
+    }
+
+    if let Some(Glyph::Composite(glyph)) = loca.get_glyf(gid, glyf).ok().flatten() {
+        for child in glyph.components() {
+            operation_count = glyf_closure_glyphs(
+                loca,
+                glyf,
+                child.glyph,
+                gids_to_retain,
+                operation_count,
+                depth,
+            );
+        }
     }
-    gids
+    operation_count
+}
+
+fn remove_invalid_gids(gids: &mut IntSet<GlyphId>, num_glyphs: usize) {
+    gids.remove_range(GlyphId::new(num_glyphs as u16)..=GlyphId::new(CODEPOINT_INVALID as u16));
+}
+
+fn get_font_num_glyphs(font: &FontRef) -> usize {
+    let mut ret = 0;
+    let loca = font.loca(None).expect("Error reading loca table");
```

Comment:
do we want to panic if these tables are missing, or should we return an error? (here and elsewhere)

Reply:
uh.. I'll work on this later in another PR, but I think we should not panic here.